src/fsharp/Optimizer.fs

```diff
@@ -3195,6 +3240,104 @@ and OptimizeApplication cenv env (f0, f0ty, tyargs, args, m) =
                    MightMakeCriticalTailcall = mayBeCriticalTailcall
                    Info=ValueOfExpr newExpr }
     
+/// Extract a sequence of pipe-right operations (note the pipe-right operator is left-associative
+/// so we start with the full thing and descend down taking apps off the end first)
+/// The pipeline begins with a |>, ||> or |||>
+and getPipes g expr acc =
+    match expr with
+    | OpPipeRight g (resType, xExpr, fExpr, m) ->
+        getPipes g xExpr (([xExpr.Range], resType, fExpr, m) :: acc) 
+    | OpPipeRight2 g (resType, x1Expr, x2Expr, fExpr, m) ->
+        [x1Expr; x2Expr], (([x1Expr.Range; x2Expr.Range], resType, fExpr, m) :: acc)
+    | OpPipeRight3 g (resType, x1Expr, x2Expr, x3Expr, fExpr, m) ->
+        [x1Expr; x2Expr; x3Expr], (([x1Expr.Range; x2Expr.Range; x3Expr.Range], resType, fExpr, m) :: acc)
+    | _ ->
+        [expr], acc
+
+and mkDebugPoint g m e = 
+    mkThenDoSequential DebugPointAtSequential.SuppressStmt m e (mkUnit g m)
+
+/// In debug code, process a pipe-right manually to lay down the debug point for the application of the function after
+/// the evaluation of the argument, all the way down the chain.
+and OptimizeDebugPipeRights cenv env expr =
+    let g = cenv.g
+
+    env.methEnv.pipelineCount <- env.methEnv.pipelineCount + 1
+    let xs0, pipes = getPipes g expr []
+    
+    let xs0R, xs0Infos = OptimizeExprsThenConsiderSplits cenv env xs0
+    let xs0Info = CombineValueInfosUnknown xs0Infos
+
+    assert (pipes.Length > 0)
+    let pipesFront, pipeLast = List.frontAndBack pipes
+
+    // The last pipe in the chain
+    //     ... |> fLast
+    // turns into a then-do sequential, so
+    //    fLast <prev-pipe-input> thendo ()
+    // with a breakpoint on the first expression
+    let binderLast (prevInputs, prevInputInfo) =
+        let (_, _, fExpr: Expr, _) = pipeLast
+        let fRange = fExpr.Range
+        let fType = tyOfExpr g fExpr
+        let fR, finfo = OptimizeExpr cenv env fExpr
+        let app = mkApps g ((fR, fType), [], prevInputs, fRange)
+        let expr = mkDebugPoint g fRange app
+        let info = CombineValueInfosUnknown [finfo; prevInputInfo]
+        expr, info
+
+    // Mid points in the chain
+    //     ... |> fMid |> rest
+    // turn into let-binding on an intermediate pipe stage
+    //    let pipe-stage-n = fMid <prev-pipe-input> 
+    //    rest <pipe-stage-n>
+    // with a breakpoint on the binding
+    //
+    let pipesBinder =
+        List.foldBack 
+            (fun (i, (xsRange, resType, fExpr: Expr, _)) binder ->
+                let name = $"Pipe #%d{env.methEnv.pipelineCount} stage #%d{i+1}"
+                let stageVal, stageValExpr = mkLocal (List.reduce unionRanges xsRange) name resType
```

Comment:
Makes sense - you have to store the result of the value as a local if we are even to see the result when stepping. Also can become beneficial if you want to see the chain of how the value transformed over the course of the pipes.